### DISCUSSION TITLE
collections/String

### DISCUSSION THREAD

**Discussion by giscus** [sunface/rust-by-practice#236] [Book Comments]
# collections/String

Learning Rust By Practice, narrowing the gap between beginner and skilled-dev with challenging examples, exercises and projects.

https://zh.practice.rs/collections/String.html

**Comment by gladinum** [comment]
易混淆概念解析 - &str 和 String 链接失效了

**Reply by sunface** [reply]
好的，谢谢～～

**Reply by zhangyixing3** [reply]
现在又失效了。

**Comment by badpx** [comment]
第7题为何要阻止 story 自动销毁呢：
`let mut story = mem::ManuallyDrop::new(story);`

**Reply by x-tong** [reply]
因为`s`和`story`指向的是同一块内存，到达`{`的时候都会释放那里，出现**二次释放（double free）**

**Reply by x-tong** [reply]
补充：如果不阻止自动销毁的话

**Reply by 17998384** [reply]
应该是到达}的时候吧

**Comment by Bruce012356** [comment]
mark

**Comment by cddsgtc** [comment]
再回首。。。

**Comment by suncunqian** [comment]
这里有人能帮我解释一下这个问题吗？谢谢了
第七题的答案中
```
let ptr = story.as_mut_ptr();
```
我看了下as_mut_ptr的函数签名：
```
pub fn as_mut_ptr(&mut self) -> *mut u8 {
    self as *mut str as *mut u8
}
```
story明明是个不可变变量， 为什么可以作为可变引用传入方法？

**Reply by dodoto** [reply]
执行这个函数之前不是已经转移到可变变量上了吗

**Comment by hexWars** [comment]
第三题是String::from("hello, world");一次然后to_string();一次吗

**Reply by CamaroW** [reply]
ChatGPT:

在这段代码中，共发生 两次 堆内存分配。让我们分析每个关键步骤：

String::from("hello, world!")

这行代码创建了一个 String，并将 "hello, world!" 这个字符串的内容存储在 堆 上。
第一次堆分配发生在这里。
let slice: &str = &s;

这里 slice 只是对 s 的借用，没有发生新的堆分配。
let s: String = slice.to_string();

slice.to_string() 会创建一个新的 String，这需要在 堆 上分配新的内存来存储 "hello, world!"。
第二次堆分配发生在这里。
assert_eq!(s, "hello, world!"); 和 println!("Success!");

这些操作不会引起额外的堆分配。
结论：
堆内存分配发生了两次：

String::from("hello, world!")
slice.to_string()

**Comment by XU0618** [comment]
done lala~

**Comment by chenhai998** [comment]
作者大大,本节页面的中英文跳转有问题,点击页面右上角按钮进行中英文页面间相互跳转,会跳到404页面.已提交pr: https://github.com/sunface/rust-by-practice/pull/442

**Comment by chenhai998** [comment]
done

**Comment by kurihada** [comment]
第三题共发生了两次堆内存分配。
1.在 let s: String = String::from("hello, world!"); 这一行，通过 String::from 函数，将 &str 类型的字符串字面量转换为 String 类型。这会在堆上分配内存来存储字符串 "hello, world!"。
2.在 let s: String = slice.to_string(); 这一行，通过调用 to_string 方法，将 &str 类型的切片 slice 转换为一个新的 String 类型。这同样会在堆上分配内存来存储新的字符串 "hello, world!"。

**Comment by hzzou** [comment]
这个的学习页面怎么没有了？只有动态数组和HashMap

**Comment by Muzych** [comment]
Mark

**Comment by felipe300** [comment]
- [x] done

**Comment by ldd19920711** [comment]
mark

**Comment by MTDickens** [comment]
done

**Comment by DylanJinx** [comment]
第五题：

`String::from_utf8` 函数返回的确实是一个 `Result<String, Utf8Error>` 类型。这里有两个可能的结果：

1. **Ok(String)**：如果提供的字节数组是有效的 UTF-8 编码，函数会返回 `Result` 的 `Ok` 变体，其中包含了构建出的 `String`。

2. **Err(Utf8Error)**：如果字节数组不是有效的 UTF-8 编码，函数会返回 `Result` 的 `Err` 变体，其中包含了一个 `Utf8Error` 类型的错误，描述了具体的解码问题。

这个设计允许调用者通过匹配 `Result` 来优雅地处理成功和错误情况。例如：

```rust
let v = vec![104, 101, 108, 108, 111]; // 字节代表 "hello"
let result = String::from_utf8(v);

match result {
    Ok(string) => println!("Converted string: {}", string),
    Err(e) => println!("Failed to convert: {}", e),
}
```

使用 `.unwrap()` 是一种快速处理 `Result` 的方法，它在 `Result` 是 `Ok` 时直接提取值，在 `Result` 是 `Err` 时引发 panic。这种方式在原型设计或者确定不会出错的场景中非常有用，但在生产代码中通常建议更仔细地处理错误。

**Reply by konlon-ya** [reply]
学到了

**Comment by DylanJinx** [comment]
done~~~

**Comment by TDAkory** [comment]
mark finished

**Comment by yangyue2001** [comment]
为什么这里的练习题大量的知识都没有提到啊

**Comment by bansal-harsh-2504** [comment]
Day 9
Done✅

**Comment by gastrodia** [comment]
第五题啊啊啊啊
```rust
fn main() {
    let mut s = String::new();
    s += "hello";

    let v = vec![104, 101, 108, 108, 111];

    // 将字节数组转换成 String
    let Ok(s1) = String::from_utf8(v) else { todo!() };
    println!("{}",s1);

    assert_eq!(s, s1);

    println!("Success!")
}
```

**Comment by Shade2012** [comment]
at first i kinda confuse but after this test it start to make sense on how the str and String work the Utf8 test is like a cherry on top. nice work
